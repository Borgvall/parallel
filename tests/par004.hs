-- Checks if runEvalIO sequence the evaluation with the IO actions as expected.

import Control.Parallel.Strategies
import Control.Concurrent
import System.IO.Unsafe

longrunning :: Int -> IO ()
longrunning decisecs = do
  threadDelay $ 10^(5::Int) * decisecs
  putStr ("Thread has been delayed by " ++ shows decisecs "/10 seconds.\n")

interleavedPrint :: String -> ()
interleavedPrint = unsafePerformIO . putStrLn

main :: IO ()
main = do
  -- check runEvalIO . r0
  lastused <- runEvalIO . r0 
                $ interleavedPrint "Should be printed at the end."
  _ <- runEvalIO . r0 $ interleavedPrint "Should never be printed."

  -- check runEvalIO . rseq
  _ <- runEvalIO . rseq $ interleavedPrint "Should be printed first."

  -- check runEvalIO . rpar
  --
  -- When using the threaded runtime the spark should print its text before
  -- the main thread. If the unthreaded runtime is used it should never be
  -- printed.
  unit <- runEvalIO . rpar . unsafePerformIO $ longrunning 1
  longrunning 2
  return unit -- make sure unit is not garbage collected

  return $! lastused -- force printing now
