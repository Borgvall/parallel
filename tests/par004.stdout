Compiling
Running
Should be printed first.
Thread has been delayed by 1/10 seconds.
Thread has been delayed by 2/10 seconds.
Should be printed at the end.
Done
